### DISCUSSION TITLE
After load of `~/Library/Application Support/Rectangle/RectangleConfig.json` JSON is moved?

### DISCUSSION THREAD

**Discussion by magnetikonline** [rxhanson/Rectangle#893] [General]
- [ ] This is a new bug that hasn't been documented in a closed issue or in the Readme.

macOS version: `12.5.1 (21G83)`
Rectangle version: v0.58
Logs if applicable (In Rectangle menu, hold option, "View Logging..."): 

I thought this was a bug:

- Exported my config to a JSON file.
- Moved into `~/Library/Application Support/Rectangle/RectangleConfig.json`
- Restarted Rectangle, config is loaded 👍
- .... but, then the JSON is renamed?

Felt this was a bug, but looking at the code - it's intended design? 

https://github.com/rxhanson/Rectangle/blob/bf967a7e05d47d6edfeb93f23c93c7b7e2d0d244/Rectangle/PrefsWindow/Config.swift#L74-L96

Problem is - another restart of Rectangle, my JSON config is no longer found/loaded.

Maybe it's just me, but this seems the wrong way around and/or not actually required?

**Comment by rxhanson** [comment]
This is to avoid writing the json out every time a change to the config is made. I wanted to get a first cut out the door, and if anyone wants to take it a step further, there's already a good starting point.

**Comment by magnetikonline** [comment]
Thanks @rxhanson - I didn't realise the config was then written _back_ to this location.

Just found the option pretty handy - I can slap my ideal Rectangle config into my [`dotfiles`](https://github.com/magnetikonline/dotfiles-macos/blob/main/app/rectangle/RectangleConfig.json) and then symlink it into place - which I do for all my "dotfiles" files/configs via a small init bash script.

Makes setting up a new system/keeping it "how I like it" really nice. This feature (for me) - is _almost_ there. Honestly, if we just deleted the lines I've marked above - I'd be 100% happy 😄. Just need the "read" of the config, no need to write/update.

**Reply by rxhanson** [reply]
Without the write functionality (that has not been added at this time - it would be a little complex) and without renaming or removing the file, you'd end up with a scenario where any changes you made to the config in the app wouldn't persist on restart - which would feel pretty broken to me. Another simple alternative that I hadn't thought of before would be to check the modified date of the file, and persist that date to check against each load. I'll give that a whirl next time (or anyone else can give that a go).

**Reply by rxhanson** [reply]
Another thing to clarify here: the load from the config file should only be necessary once - because those values will be written to NSUserDefaults, and they shouldn't need to be loaded again. If for some reason you don't see those values persisted on restart, then there is a bug.

**Reply by magnetikonline** [reply]
> Without the write functionality and without renaming or removing the file, you'd end up with a scenario where any changes you made to the config in the app wouldn't persist on restart

Ah fair call.

> Another simple alternative that I hadn't thought of before would be to check the modified date of the file, and persist that date to check against each load.

That's not bad. Maybe even _easier_ - could allow for a _second_ config filename - e.g. `RectangleConfig.Fixed.json` (or something better named) - this is checked _first_ - if found, then it's always loaded - and that's it - no renames/further searching? Or maybe that's a big ugly...?

**Reply by magnetikonline** [reply]
> If for some reason you don't see those values persisted on restart, then there is a bug.

No that's very much working as you've outlined. I just saw the config more as like a `.gitconfig` - lives on disk and is _always_ the authoritative source of truth. Realise now that probably not the entire intention. 👍

**Reply by rxhanson** [reply]
Actually, I would like for that to be the way it works, but it would have taken a much larger time investment that I wasn't ready to pour into it right now. I was pointed to the way iTerm2 does it, which is quite nice. Someday...

**Reply by magnetikonline** [reply]
Funny - I was going to mention iTerm2, but yeah - there is quite a bit more UI/logic to add in there to hit that parity. 👍

**Comment by magnetikonline** [comment]
And slightly off topic @rxhanson - love the new "hot regions" actions UI in the settings pane 👍.
